### PR TITLE
Add public free method

### LINKED ISSUE #12: Add manual/public Free method
https://twitter.com/dgryski/status/981666686741827584

### PR DESCRIPTION
**Pull request by Oppen**
Resolves #12

### CODE REVIEW

**Comment by lemire** [comment]
We seem to have CI errors...

File: free_test/explicit_free.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	bitmaps = nil
+	runtime.GC()
+	fmt.Printf("Copied then GC'd %d bitmaps... ", N)
+	if err := syscall.Getrusage(syscall.RUSAGE_SELF, &rusage); err != nil {
+		fmt.Printf("Getrusage failed with error: %s\n", err)
+		os.Exit(1)
+	}
+	fmt.Printf("max RSS: %d bytes\n", rusage.Maxrss)
+}
```
Comment (willglynn) [inline free_test/explicit_free.go:102]: Should this test compare `rusage.Maxrss` between scenarios?

Comment (Oppen) [reply free_test/explicit_free.go:102]: Maybe. I guess checking it grows at least 100x is reasonable.

**Comment by Oppen** [comment]
The jobs appear as cancelled, which seems odd. Maybe a timeout? Maybe I should decrease `N` a bit?

File: free_test/explicit_free.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		fmt.Printf("Getrusage failed with error: %s\n", err)
+		os.Exit(1)
+	}
+	fmt.Printf("max RSS: %d bytes\n", rusage.Maxrss)
+
+	for i := range bitmaps {
+		bitmaps[i] = bitmap.Clone()
+		bitmaps[i].Free()
+	}
+	fmt.Println("Copied then explicitly freed %d bitmaps... ")
```
Comment (Oppen) [inline free_test/explicit_free.go]:
```suggestion
	fmt.Printf("Copied then explicitly freed %d bitmaps... ", N)
```

**Comment by lemire** [comment]
@Oppen 

I don't think that they are cancelled. I just think that we must have a fast-fail and that they stop after the initial failure...

```
CRoaring 0.2.65
Alloc = 4 MiB	TotalAlloc = 41 MiB	Sys = 70 MiB	NumGC = 12
Alloc = 3 MiB	TotalAlloc = 80 MiB	Sys = 70 MiB	NumGC = 23
Alloc = 4 MiB	TotalAlloc = 120 MiB	Sys = 70 MiB	NumGC = 34
Alloc = 3 MiB	TotalAlloc = 159 MiB	Sys = 70 MiB	NumGC = 46
Alloc = 4 MiB	TotalAlloc = 198 MiB	Sys = 70 MiB	NumGC = 57
Alloc = 3 MiB	TotalAlloc = 238 MiB	Sys = 70 MiB	NumGC = 69
Alloc = 4 MiB	TotalAlloc = 277 MiB	Sys = 70 MiB	NumGC = 81
Alloc = 3 MiB	TotalAlloc = 316 MiB	Sys = 70 MiB	NumGC = 92
Alloc = 4 MiB	TotalAlloc = 356 MiB	Sys = 70 MiB	NumGC = 104
Alloc = 3 MiB	TotalAlloc = 395 MiB	Sys = 70 MiB	NumGC = 115
ok
size in kB of one bitmap  1224 ; number of copies =  7 ; total alloc:  8573 kB
TestExplicitFree output:
Starting execution... max RSS: 25636 bytes
Created reference bitmap... max RSS: 25636 bytes
Allocated slice for 100000 bitmap pointers... max RSS: 25636 bytes
Copied then explicitly freed 100000 bitmaps... max RSS: 25636 bytes
GC'd old bitmaps... max RSS: 25636 bytes
Allocated slice for 100000 bitmap pointers... max RSS: 25636 bytes

--- FAIL: TestExplicitFree (28.74s)
```


Please note that nothing in the C specification allows you to believe that `free` will return the memory to the operating system.

**Comment by lemire** [comment]
However, if you constantly allocate and call free, many many times... the memory usage should not grow without bounds. (I don't know if that's required by the C standard, but practice dictates it.)

**Comment by Oppen** [comment]
> Please note that nothing in the C specification allows you to believe that free will return the memory to the operating system.

I'm aware. But the closest we can get to measure frees without either a custom `malloc` that provides stats such as `tcmalloc` or running valgrind and somehow aborting before collection is to just hope the allocator recycles memory you gave back to it, thus not asking for more (i.e., not seeing it growing linearly is a good sign). I know it relies on a lot of assumptions, including that it will always be able to ask for memory from the same lists.

Maybe the test is not finding the `go` binary in it `PATH`?

**Comment by lemire** [comment]
> Maybe the test is not finding the go binary in it PATH?

Given that the script calls `go test something`, if `go` was not on `PATH`, then nothing would work at all.

We do see the `TestExplicitFree output:` output which suggests that the call goes through.

**Comment by lemire** [comment]
This is the error...

```
--- FAIL: TestExplicitFree (28.74s)
28
    gocroaring_test.go:64: Command exited with error: exit status 1
```

**Comment by lemire** [comment]
I ran the test locally and my machine was burning up...

```
❯ go test
CRoaring 0.2.65
Alloc = 4 MiB	TotalAlloc = 41 MiB	Sys = 70 MiB	NumGC = 11
Alloc = 3 MiB	TotalAlloc = 80 MiB	Sys = 70 MiB	NumGC = 23
Alloc = 2 MiB	TotalAlloc = 120 MiB	Sys = 70 MiB	NumGC = 35
Alloc = 4 MiB	TotalAlloc = 159 MiB	Sys = 70 MiB	NumGC = 46
Alloc = 2 MiB	TotalAlloc = 199 MiB	Sys = 70 MiB	NumGC = 58
Alloc = 4 MiB	TotalAlloc = 238 MiB	Sys = 70 MiB	NumGC = 69
Alloc = 2 MiB	TotalAlloc = 277 MiB	Sys = 70 MiB	NumGC = 81
Alloc = 4 MiB	TotalAlloc = 317 MiB	Sys = 70 MiB	NumGC = 92
Alloc = 2 MiB	TotalAlloc = 356 MiB	Sys = 70 MiB	NumGC = 104
Alloc = 4 MiB	TotalAlloc = 395 MiB	Sys = 70 MiB	NumGC = 115
ok
size in kB of one bitmap  1224 ; number of copies =  7 ; total alloc:  8573 kB
TestExplicitFree output:
Starting execution... max RSS: 15712256 bytes
Created reference bitmap... max RSS: 18448384 bytes
Allocated slice for 100000 bitmap pointers... max RSS: 18677760 bytes
Copied then explicitly freed 100000 bitmaps... max RSS: 27590656 bytes
GC'd old bitmaps... max RSS: 30900224 bytes
Allocated slice for 100000 bitmap pointers... max RSS: 30916608 bytes
Copied then GC'd 100000 bitmaps... max RSS: 10714710016 bytes
```

I think it fails because it can't allocate that much memory.

**Comment by lemire** [comment]
In CI, we have limited memory... certainly not gigabytes... 

A few kilobytes of allocated memory should be enough, I would think.

(I have manually checked that there does not seem to be memory leaks in the sense that allocated bitmaps do get freed...)

**Comment by Oppen** [comment]
Yeah, maybe I was a bit extreme to make sure the number grows big. I'll be more conservative.

**Comment by lemire** [comment]
I am now happy to merge this if you are happy.

**Comment by Oppen** [comment]
I am happy.

**Comment by lemire** [comment]
Merging.

**Pull request merged**